src/goto-programs/json_goto_trace.cpp

```diff
@@ -11,20 +11,157 @@ Author: Daniel Kroening
 #include <cassert>
 
 #include <util/json_expr.h>
+#include <util/arith_tools.h>
 
 #include <langapi/language_util.h>
 
 #include "json_goto_trace.h"
 
 /*******************************************************************\
 
+Function: remove_pointer_offsets
+
+  Inputs:
+    src - an expression
+
+ Purpose: Replaces in src, expressions of the form pointer_offset(constant)
+          by that constant.
+
+\*******************************************************************/
+
+void remove_pointer_offsets(exprt &src)
+{
+  if(src.id()==ID_pointer_offset && src.op0().id()==ID_constant)
+    src=src.op0();
+  else
+    for(auto &op : src.operands())
+      remove_pointer_offsets(op);
+}
+
+/*******************************************************************\
+
+Function: remove_pointer_offsets
+
+  Inputs:
+    src - an expression
+    array_symbol - a symbol expression representing an array
+
+ Purpose: Replaces in src, expressions of the form
+          pointer_offset(array_symbol) by a constant value of 0.
+          This is meant to simplify array expressions.
+
+\*******************************************************************/
+
+void remove_pointer_offsets(exprt &src, symbol_exprt array_symbol)
+{
+  if(src.id()==ID_pointer_offset &&
+     src.op0().id()==ID_constant &&
+     src.op0().op0().id()==ID_address_of &&
+     src.op0().op0().op0().id()==ID_index)
+  {
+    const index_exprt &idx=to_index_expr(src.op0().op0().op0());
+    const irep_idt &array_id=to_symbol_expr(idx.array()).get_identifier();
+    if(idx.array().id()==ID_symbol &&
+       array_id==array_symbol.get_identifier() &&
+       to_constant_expr(idx.index()).value_is_zero_string())
+      src=from_integer(0, src.type());
+  }
+  else
+    for(auto &op : src.operands())
+      remove_pointer_offsets(op, array_symbol);
+}
+
+/*******************************************************************\
+
+Function: simplify_index
+
+  Inputs:
+    idx - an expression representing an index in an array
+    out - a reference to an unsigned value
+
+ Outputs: Boolean flag that is true if the `idx` expression could not be
+          simplified into a unsigned constant value.
+
+ Purpose: Simplify the expression in index as much as possible to try
+          to get a unsigned constannt.
+
+\*******************************************************************/
+
+bool simplify_index(const exprt &idx, unsigned &out)
+{
+  if(idx.id()==ID_constant)
+  {
+    std::string value_str=id2string(to_constant_expr(idx).get_value());
+    mp_integer int_value=binary2integer(value_str, false);
+    out=int_value.to_long();
+    return false;
+  }
+  else if(idx.id()==ID_plus)
```

Comment:
All examples I have seen are of the form a pointer offset address plus some constant index, but I admit I don't know whether that really covers all the possibilities.